scripts/speech_recognition/estimate_duration_bins_2d.py

```diff
@@ -0,0 +1,319 @@
+    )
+    parser.add_argument("--text-field", default="text", help="The key in manifests to read transcripts from.")
+    parser.add_argument("--lang-field", default="lang", help="The key in manifests to read language from.")
+    parser.add_argument(
+        "-n",
+        "--num_examples",
+        type=int,
+        default=-1,
+        help="The number of examples (utterances) to estimate the bins. -1 means use all data "
+        "(be careful: it could be iterated over infinitely).",
+    )
+    parser.add_argument(
+        "-l",
+        "--min_duration",
+        type=float,
+        default=-float("inf"),
+        help="If specified, we'll filter out utterances shorter than this.",
+    )
+    parser.add_argument(
+        "-u",
+        "--max_duration",
+        type=float,
+        default=float("inf"),
+        help="If specified, we'll filter out utterances longer than this.",
+    )
+    parser.add_argument(
+        "--max_tps",
+        type=float,
+        default=float("inf"),
+        help="If specified, we'll filter out utterances with more tokens/second than this. "
+        "On regular utterances and BPE tokenizers with 1024 tokens 10-12tps is generally a reasonable limit.",
+    )
+    parser.add_argument(
+        "-q", "--quiet", type=bool, default=False, help="When specified, only print the estimated duration bins."
+    )
+    parser.add_argument(
+        "-f",
+        "--prompt-format",
+        type=str,
+        help="When specified, we'll use a prompt formatter in addition to the tokenizer for the purpose of estimating token count bins. "
+        "This is useful for accurate 2D bucket estimation with models such as EncDecMultiTaskModel (Canary-1B), "
+        "or any model where the label sequence consists of a user prompt and a model's response.",
+    )
+    parser.add_argument(
+        "-p",
+        "--prompt",
+        type=str,
+        help="Prompt slots provided as a Python list of dicts. It is used together with --prompt-format option."
+        "For example, with Canary-1B you may use: [{'role':'user','slots':{'source_lang':'en','target_lang':'en','task':'asr','pnc':'yes'}]",
+    )
+    return parser.parse_args()
+
+
+def estimate_duration_buckets(
+    cuts: Iterable[Cut],
+    num_buckets: int,
+    num_subbuckets: int,
+    max_tps: float,
+    max_duration: float,
+    quiet: bool,
+) -> list[tuple[float, float]]:
+    assert num_buckets > 1
+
+    constraint = FixedBucketBatchSizeConstraint2D([(0.0, 0.0)], [0])
+
+    sizes = []
+    num_tokens = []
+    for c in cuts:
+        dur, toks = constraint.measure_length(c)
+        sizes.append(dur)
+        num_tokens.append(toks)
+    sizes = np.array(sizes, dtype=np.float32)
+    num_tokens = np.array(num_tokens, dtype=np.int32)
+    joint = np.rec.fromarrays([sizes, num_tokens])
+    joint.sort()
+    sizes = joint.f0
+    num_tokens = joint.f1
+
+    size_per_bucket = sizes.sum() / num_buckets
```

Comment:
I don't know if you ever said this explicitly, but it took me a while to understand that you are choosing the bucket lower and upper bounds on duration such that every bucket has approximately the same duration.

Reply:
That's right. This is the relevant discussion about it (although a lot of other facts mentioned there about lhotse are already outdated, as is the sampler we used there itself) https://github.com/lhotse-speech/lhotse/issues/364